### PR TITLE
Meeting plugin

### LINKED ISSUE #40: Implement initial collaborative features

### PR DESCRIPTION
**Pull request by sumboid**
Resolves #40

### CODE REVIEW

File: packages/webrtc/src/api.ts [hunk 1]
```diff
@@ -0,0 +1,179 @@
+//
```
Comment (haiodo) [inline packages/webrtc/src/api.ts:1]: @sumboid Could you please check #55, should we have some unified json RPC stuff for both server APIs.

File: packages/webrtc/src/client.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import { ReqID, Incoming, Outgoing, Request, NotificationMethod, IncomingNotifications, OutgoingNotifications } from './api'
+
+function serialize (object: any): string {
```
Comment (haiodo) [inline packages/webrtc/src/client.ts:18]: Could you please check #55, could we use some similar/same RPC stuff?

File: models/meeting/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,13 @@
+{
+  "compilerOptions": {
+    "target": "es6",
+    "module": "commonjs",
+    "declaration": true,
+    "declarationMap": true,
+    "outDir": "./lib",
+    "rootDir": "./src",
+    "strict": true,
+    "experimentalDecorators": true,
```
[the remark this answers is not in the document]
Comment (sumboid) [reply models/meeting/tsconfig.json:10]: Yep, right here to define `Room` space: https://github.com/hardcoreeng/platform/pull/60/files#diff-7ac252dd0ee994dbf0f84aaefeeaf848ecc11e4537a7c4d8fd5709bcc2793f16R24

File: packages/webrtc/src/api.ts (hunk 1 above)
Comment (sumboid) [reply packages/webrtc/src/api.ts:1]: Yeah I had a thought about that, let's discuss this on meeting: I think to use same client we need to rework `webrtc-service` a bit

File: common/config/rush/.pnpmfile.cjs [hunk 4]
```diff
@@ -26,13 +26,21 @@ module.exports = {
...
 
-  return packageJson;
+  // Installation script of ws 0.4.32 does some weird things as the result other
+  // packages could be affected by its output, specifically @babel/types
+  // This ws version is kurento-client dependency
+  if (pkg.name === 'websocket-stream' && pkg.version.startsWith('0.')) {
+    context.log('Fix outdated ws dependency version')
+    pkg.dependencies.ws = '1.1.5';
+  }
+
```
Comment (z2sx) [inline common/config/rush/.pnpmfile.cjs:44]: Why this file is different? Different `rush`/`pnpm` versions?

File: packages/webrtc/src/api.ts [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+const screenSuffix = '-screen'
+export const makeScreenID = (id: InternalID): InternalID => `${id}${screenSuffix}`
+export const getScreenOwner = (id: InternalID): InternalID => id.slice(0, -screenSuffix.length)
+export const isScreenID = (id: InternalID): boolean => id.endsWith(screenSuffix)
+
+interface BaseReq {
+  id: ReqID
+  method: ReqMethod
+}
+
```
Comment (z2sx) [inline packages/webrtc/src/api.ts:52]: I think this and below should use `@anticrm/rpc` stuff

File: packages/webrtc/src/client.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+function deserialize<D extends Incoming | Outgoing> (response: string): D {
+  return JSON.parse(response)
+}
+
+type Subscriber = (x: OutgoingNotifications) => Promise<void> | void
+
+export class Client {
+  private reqId: number = 0
+  private readonly pending = new Map<ReqID, (x: Outgoing) => void>()
```
Comment (z2sx) [inline packages/webrtc/src/client.ts:30]: Is this stuff similar to `RequestProcessor`?

File: plugins/meeting/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+  screen: Readable<Peer>
+  status: Readable<Status>
+}
+
+export interface MeetingService extends Service {
+  room: Room
+  join: (room: string) => Promise<void>
+  leave: () => Promise<void>
+  shareScreen: () => Promise<void>
+  stopSharingScreen: () => Promise<void>
```
Comment (z2sx) [inline plugins/meeting/src/index.ts:46]: `Service` may become obsolete in future versions, do you really need this methods available to other plugins?

File: common/config/rush/.pnpmfile.cjs (hunk 4 above)
Comment (sumboid) [reply common/config/rush/.pnpmfile.cjs:44]: Do I understand correctly that we have `rush` over `pnpm`? In general what we do here is changing internal dependency of `websocket-stream` on the fly, this happens during `rush update`.

File: packages/webrtc/src/client.ts (hunk 6 above)
Comment (sumboid) [reply packages/webrtc/src/client.ts:30]: Yep, as we discussed, it was simplified to match `webrtc-service` capabilities, I plan to migrate to `@anticrm/rpc` as soon as #55 be merged

File: plugins/meeting/src/index.ts (hunk 7 above)
Comment (sumboid) [reply plugins/meeting/src/index.ts]: Yeah, good point, made all this stuff internal until we get some real cases

**Comment by sumboid** [comment]
@haiodo @z2sx meeting service was moved to separate repo: https://github.com/hardcoreeng/meeting-service/pull/1
so wrt pnpm hack, it is not required anymore

**Pull request merged**